File: ios/Touliao/Features/Chat/HistoryPaginationAction.swift
import Foundation

protocol HistoryPageSource {
    func loadHistory(
        _ conversationId: String,
        before: Double?,
        beforeId: String?
    ) async throws -> [Message]
}

struct HistoryPaginationState {
    var messages: [Message]
    var loadingEarlier: Bool = false
    var reachedStart: Bool = false
}

@MainActor
struct HistoryPaginationAction {
    let source: HistoryPageSource
    let pageSize: Int

    init(source: HistoryPageSource, pageSize: Int = 50) {
        self.source = source
        self.pageSize = pageSize
    }

    func execute(
        conversationId: String,
        state: () -> HistoryPaginationState,
        isCurrentAttempt: () -> Bool,
        apply: (HistoryPaginationState) -> Void
    ) async {
        guard isCurrentAttempt() else { return }
        var current = state()
        guard !current.loadingEarlier, !current.reachedStart,
              let boundary = current.messages.first(where: { $0.localStatus == nil && !$0.id.isEmpty }) else { return }

        current.loadingEarlier = true
        apply(current)
        defer {
            if isCurrentAttempt() {
                var latest = state()
                latest.loadingEarlier = false
                apply(latest)
            }
        }

        guard let older = try? await source.loadHistory(
            conversationId,
            before: boundary.createdAt,
            beforeId: boundary.id
        ), isCurrentAttempt() else { return }
        current = state()
        let existing = Set(current.messages.map(\.id))
        current.messages = older.filter { !existing.contains($0.id) } + current.messages
        current.reachedStart = older.count < pageSize
        apply(current)
    }
}

File: android/app/src/main/java/com/touliao/app/feature/chat/HistoryPaginationAction.kt
package com.touliao.app.feature.chat

import com.touliao.app.data.repository.HistoryPageSource
import kotlinx.coroutines.CancellationException

internal class HistoryPaginationAction(
    private val source: HistoryPageSource,
    private val pageSize: Int = 50,
) {
    suspend fun execute(
        conversationId: String,
        state: () -> ChatUiState,
        isCurrentAttempt: () -> Boolean,
        updateState: (((ChatUiState) -> ChatUiState)) -> Unit,
    ) {
        if (!isCurrentAttempt()) return
        val snapshot = state()
        if (snapshot.loadingEarlier || snapshot.reachedStart) return
        val boundary = snapshot.messages.firstOrNull {
            it.localStatus == null && it.id.isNotBlank()
        } ?: return

        var started = false
        updateState { current ->
            if (!isCurrentAttempt() || current.loadingEarlier || current.reachedStart) current
            else current.copy(loadingEarlier = true).also { started = true }
        }
        if (!started) return

        try {
            val older = source.loadHistory(
                conversationId = conversationId,
                before = boundary.created_at,
                beforeId = boundary.id,
            )
            if (!isCurrentAttempt()) return
            updateState { current ->
                if (!isCurrentAttempt()) current
                else {
                    val existing = current.messages.mapTo(HashSet()) { it.id }
                    current.copy(
                        messages = older.filterNot { it.id in existing } + current.messages,
                        reachedStart = older.size < pageSize,
                    )
                }
            }
        } catch (error: Exception) {
            if (error is CancellationException) throw error
        } finally {
            if (isCurrentAttempt()) {
                updateState { current ->
                    if (isCurrentAttempt()) current.copy(loadingEarlier = false) else current
                }
            }
        }
    }
}
